pkg/engine/plugins.go

```diff
@@ -183,3 +191,95 @@ func installPlugin(client deploy.BackendClient, plugin workspace.PluginInfo) err
+// reflecting the version of a provider that should be used as the "default" resource when registering resources. This
+// function takes two sets of plugins: a set of plugins given to us from the language host and the full set of plugins.
+// If the language host has sent us a non-empty set of plugins, we will use those exclusively to service default
+// provider requests. Otherwise, we will use the full set of plugins, which is the existing behavior today.
+//
+// The justification for favoring language plugins over all else is that, ultimately, it is the language plugin that
+// produces resource registrations and therefore it is the language plugin that should dictate exactly what plugins to
+// use to satisfy a resource registration. Since we do not today request a particular version of a plugin via
+// RegisterResource (pulumi/pulumi#2389), this is the best we can do to infer the version that the language plugin
+// actually wants.
+//
+// Whenever a resource arrives via RegisterResource and does not explicitly specify which provider to use, the engine
+// injects a "default" provider resource that will serve as that resource's provider. This function computes the map
+// that the engine uses to determine which version of a particular provider to load.
+//
+// it is critical that this function be 100% deterministic.
+func computeDefaultProviderPlugins(languagePlugins, allPlugins pluginSet) map[tokens.Package]*semver.Version {
+	// Language hosts are not required to specify the full set of plugins they depend on. If the set of plugins received
+	// from the language host is the empty set, fall back to using the list of all plugins instead.
+	sourceSet := languagePlugins
+	if sourceSet.Empty() {
+		logging.V(preparePluginLog).Infoln(
+			"computeDefaultProviderPlugins(): language host reported empty set of plugins, using all plugins")
+		sourceSet = allPlugins
+	}
+
+	defaultProviderVersions := make(map[tokens.Package]*semver.Version)
+
+	// Sort the set of source plugins by version, so that we iterate over the set of plugins in a deterministic order.
+	// Sorting by version gets us two properties:
+	//   1. The below loop will never see a nil-versioned plugin after a non-nil versioned plugin, since the sort always
+	//      considers nil-versioned plugins to be less than non-nil versioned plugins.
+	//   2. The below loop will never see a plugin with a version that is older than a plugin that has already been
+	//      seen. The sort will always have placed the older plugin before the newer plugin.
+	//
+	// Despite these properties, the below loop explicitly handles those cases to preserve correct behavior even if the
+	// sort is not functioning properly.
+	sourcePlugins := sourceSet.Values()
+	sort.Sort(workspace.SortedPluginInfo(sourcePlugins))
+	for _, p := range sourcePlugins {
+		logging.V(preparePluginLog).Infof("computeDefaultProviderPlugins(): considering %s", p)
+		if p.Kind != workspace.ResourcePlugin {
+			continue
```

Comment:
consider both logging here, as well as having comment explaining what this is for.